File: question/neetcode150/sliding_window/minimum-window-substring.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    string minWindow(string t, string s) {
        int l = 0;
        int al = 0, ar=INT_MAX;

        int n1 = s.size();
        int n2 = t.size();

        unordered_map<char, int> d1;
        for (auto c:s){
            d1[c]++;
        }

        unordered_map<char, int> d2;

        int have = 0, need = d1.size();
        
        for (int r=0; r<n2; r++){
            char c = t[r];
            d2[c]++;

            if(d1.count(c) and d2[c]==d1[c]){
                have++;
            }

            while (have==need){
                if (r-l < ar-al){
                    al = l, ar = r;
                }
                char c = t[l];
                d2[c]--;
                if(d1.count(c) and d2[c] < d1[c]){
                    have--;
                }
                l++;
            }

        }

        if (ar != INT_MAX) 
            return t.substr(al,ar-al+1);
        return "";
    }
};

File: question/neetcode150/sliding_window/minimum-window-substring.py
from collections import defaultdict
import math

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        l = 0
        d1 = defaultdict(int)
        for c in t:
            d1[c] += 1

        d2 = defaultdict(int)
        al, ar = -math.inf, math.inf
        have, need = 0, len(d1)

        for r,c in enumerate(s):
            d2[c] += 1
            if c in d1 and d1[c] == d2[c]:
                have += 1
            while have >= need:
                if r-l < ar-al:
                    al,ar = l,r
                c = s[l]
                if c in d1 and d2[c] == d1[c]:
                    have -= 1
                d2[c]-=1
                l+=1

        return s[al:ar+1] if al != -math.inf else ""